File: src/ObjectInspector.cpp
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 athre0z
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "ObjectInspector.hpp"

#include <ida.hpp>
#include <kernwin.hpp>
#include <QDebug>
#include <QMouseEvent>

// ========================================================================= //
// [ObjectInspector]                                                         //
// ========================================================================= //

ObjectInspector::ObjectInspector(QWidget *parent)
    : QWidget(parent)
    , m_curSelectedWidget(nullptr)
    , m_updateKey(QKeySequence(Qt::Key_F7), this)
{
    m_updateKey.setContext(Qt::ApplicationShortcut);

    m_widgets.setupUi(this);
    QFont font("Monospace");
    font.setStyleHint(QFont::TypeWriter);
    m_widgets.teInspectionResults->setFont(font);

    connect(m_widgets.btnSelectParent, SIGNAL(released()), SLOT(selectParent()));
    connect(&m_updateKey, SIGNAL(activated()), SLOT(updateInspection()));
}

void ObjectInspector::updateInspection()
{
    auto widget = qApp->widgetAt(QCursor::pos());
    updateSelectedWidget(widget);
}

void ObjectInspector::selectParent()
{
    if (m_curSelectedWidget)
    {
        auto parent = m_curSelectedWidget->parent();
        if (parent && parent->inherits("QWidget"))
        {
            updateSelectedWidget(static_cast<QWidget*>(parent));
        }
    }
}

void ObjectInspector::onSelectedWidgetDestroyed(QObject*)
{
    m_curSelectedWidget = nullptr;
}

void ObjectInspector::updateSelectedWidget(QWidget *newWidget)
{
    if (m_curSelectedWidget)
    {
        disconnect(m_curSelectedWidget, SIGNAL(destroyed(QObject*)), 
            this, SLOT(onSelectedWidgetDestroyed(QObject*)));
    }

    m_curSelectedWidget = newWidget;

    if (newWidget)
    {
        m_widgets.btnSelectParent->setEnabled(newWidget->parent() != nullptr);
        m_widgets.teInspectionResults->setText(
            QString("Type: ") + newWidget->metaObject()->className()
            + "\nName: " + (newWidget->objectName().isEmpty() ? "<none>" : newWidget->objectName())
            + "\nNumber of children: " + QString::number(newWidget->children().length())
        );

        connect(newWidget, SIGNAL(destroyed(QObject*)), 
            SLOT(onSelectedWidgetDestroyed(QObject*)));
    }
    else
    {
        m_widgets.teInspectionResults->setText("<no object under cursor>");
    }
}

// ========================================================================= //

File: plugins/idaskins/objectinspector.py
from __future__ import absolute_import, division, print_function

import os

from idaskins import UI_DIR
from PyQt5 import uic
from PyQt5.Qt import qApp
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QCursor, QFont, QKeySequence
from PyQt5.QtWidgets import QShortcut, QWidget

Ui_ObjectInspector, ObjectInspectorBase = uic.loadUiType(
    os.path.join(UI_DIR, 'ObjectInspector.ui')
)


class ObjectInspector(ObjectInspectorBase):
    """
    Rudimentary Qt object inspector.
    Allows for easier finding of object names and classes
    for usage in QSS stylesheets.
    """
    def __init__(self, *args, **kwargs):
        super(ObjectInspector, self).__init__(*args, **kwargs)

        self._selected_widget = None
        self._ui = Ui_ObjectInspector()
        self._ui.setupUi(self)

        # Make everything monospace.
        font = QFont('Monospace')
        font.setStyleHint(QFont.TypeWriter)
        self._ui.teInspectionResults.setFont(font)

        # Register signals.
        self._update_key = QShortcut(QKeySequence(Qt.Key_F7), self)
        self._ui.btnSelectParent.released.connect(self.select_parent)
        self._update_key.activated.connect(self.update_inspection)

    def update_inspection(self):
        widget = qApp.widgetAt(QCursor.pos())
        self.update_selected_widget(widget)

    def select_parent(self):
        if self._selected_widget:
            parent = self._selected_widget.parent()
            if parent and parent.inherits('QWidget'):
                self.update_selected_widget(parent)

    def update_selected_widget(self, widget):
        if self._selected_widget:
            self._selected_widget.destroyed.disconnect(
                self.on_selected_widget_destroyed
            )

        self._selected_widget = widget

        if widget:
            self._ui.btnSelectParent.setEnabled(widget.parent() is not None)
            self._ui.teInspectionResults.setText((
                "Type: {}\n"
                "Name: {}\n"
                "Number of children: {}\n"
                "QSS: {}"
            ).format(
                widget.metaObject().className(),
                widget.objectName() or '<none>',
                len(widget.children()),
                widget.styleSheet() or '<none>',
            ))

            self._selected_widget.destroyed.connect(
                self.on_selected_widget_destroyed
            )
        else:
            self._ui.teInspectionResults.setText('<no object under cursor>')

    def on_selected_widget_destroyed(self, obj):
        self._selected_widget = None
